mars/services/scheduling/supervisor/globalresource.py

```diff
@@ -0,0 +1,202 @@
+                self._update_band_usage(band, subtask_resource)
+                idx += 1
+        if idx == 0:
+            logger.debug(
+                "No resources available, status: %r, request: %r",
+                self._band_used_resources,
+                subtask_resources,
+            )
+        return subtask_ids[:idx]
+
+    @mo.extensible
+    def update_subtask_slots(
+        self, band: BandType, session_id: str, subtask_id: str, slots: int
+    ):
+        self.update_subtask_resources(
+            band, session_id, subtask_id, Resource(num_cpus=slots)
+        )
+
+    @mo.extensible
+    def update_subtask_resources(
+        self, band: BandType, session_id: str, subtask_id: str, resource: Resource
+    ):
+        session_subtask_id = (session_id, subtask_id)
+        subtask_resources = self._band_stid_resources[band]
+        if session_subtask_id not in subtask_resources:
+            return
+
+        resource_delta = resource - subtask_resources[session_subtask_id]
+        subtask_resources[session_subtask_id] = resource
+        self._update_band_usage(band, resource_delta)
+
+    @mo.extensible
+    def release_subtask_slots(self, band: BandType, session_id: str, subtask_id: str):
+        self.release_subtask_resource(band, session_id, subtask_id)
+
+    @mo.extensible
+    def release_subtask_resource(
+        self, band: BandType, session_id: str, subtask_id: str
+    ):
+        # todo ensure slots released when subtasks ends in all means
+        resource_delta = self._band_stid_resources[band].pop(
+            (session_id, subtask_id), ZeroResource
+        )
+        self._update_band_usage(band, -resource_delta)
+
+    def _update_slot_usage(self, band: BandType, slots_usage_delta: float):
+        self._update_band_usage(band, Resource(num_cpus=slots_usage_delta))
+
+    def _update_band_usage(self, band: BandType, band_usage_delta: Resource):
+        self._band_used_resources[band] += band_usage_delta
+        # some code path doesn't call `apply_subtask_resources`
+        band_total_resource = self._band_total_resources.get(band)
+        if (
+            band_total_resource is not None
+            and self._band_used_resources[band] > band_total_resource
+        ):  # pragma: no cover
+            raise Exception(
+                f"Resource exceed: band used resource {self._band_used_resources[band]} "
+                f"band total resource {self._band_total_resources[band]}"
+            )
+        if self._band_used_resources[band] <= ZeroResource:
+            self._band_used_resources.pop(band)
+            self._band_idle_start_time[band] = time.time()
+            if band in self._band_idle_events:
+                self._band_idle_events.pop(band).set()
+        else:
+            self._band_idle_start_time[band] = -1
+
+    def get_used_slots(self) -> Dict[BandType, float]:
+        return {
+            band: resource.num_cpus
```

Comment:
For GPU bands the attribute should be `num_gpus`.

Reply:
deleted `get_used_slots` method, use `get_used_resources` instead.

Reply:
We could use `band: resource.num_cpus or resource.num_gpus`, because a subtask must run on either cpu or gpu.